### PR TITLE
Twisted client, examples, and benchmarking

### PR DESCRIPTION
**Pull request by RyanGWU82**
Again, the `benchmark` code isn't too important to review, the `twisted_client.py` and `twisted_pool.py` are more critical. Those haven't changed much from the link I sent yesterday.

### CODE REVIEW

**Comment by RyanGWU82** [comment]
@aren Gentle ping -- I'd like to merge this before the ldapserver changes.

File: twisted_poc.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+def connectionMade(divvyClient):
+    # Ten random IP addresses, to use for rate limiting examples
+    ip_addresses = [_random_ip() for _ in range(unique_ips)]
+
+    print("*** Connected to Divvy. Enqueueing {} checks.".format(
+        request_count))
+    for i in range(request_count):
+        client_ip = random.choice(ip_addresses)
+        hit_args = {"type": "ldap_login", "ip": client_ip}
+        resp = divvyClient.checkRateLimit(**hit_args)
```
Comment (aren) [inline twisted_poc.py]: `d =`, just to follow convention

File: twisted_pool_poc.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    ip_addresses = [_random_ip() for _ in range(unique_ips)]
+
+    addr = IPv4Address('TCP', '52.41.9.85', 8321)
+    pool = DivvyPool(addr, maxClients=connections)
+    print("*** Initialized pool with {} conns; enqueueing {} checks.".format(
+        connections, request_count))
+    for i in range(request_count):
+        client_ip = random.choice(ip_addresses)
+        hit_args = {"type": "ldap_login", "ip": client_ip}
+        resp = pool.checkRateLimit(**hit_args)
```
Comment (aren) [inline twisted_pool_poc.py]: `d =`

File: twisted_pool_poc.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+def connectionFailed(f):
+    print("*** connectionFailed:")
+    print("   f = {}".format(f))
+
+
+def main():
+    # Ten random IP addresses, to use for rate limiting examples
+    ip_addresses = [_random_ip() for _ in range(unique_ips)]
+
+    addr = IPv4Address('TCP', '52.41.9.85', 8321)
```
Comment (aren) [inline twisted_pool_poc.py]: maybe move this to a variable at the top of the file? `host = ('52.41.9.85', 8321)`  or some-such?

File: twisted_poc.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        resp.addCallback(_cbRateLimit, client_ip).addErrback(_cbDivvyError)
+
+
+def connectionFailed(f):
+    print("*** connectionFailed:")
+    print("   f = {}".format(f))
+
+
+def main():
+    point = TCP4ClientEndpoint(reactor, "52.41.9.85", 8321)
```
Comment (aren) [inline twisted_poc.py]: maybe move this to a variable at the top of the file? host = ('52.41.9.85', 8321) or some-such?

File: divvy/client.py [hunk 5]
```diff
@@ -28,12 +28,23 @@ def __init__(self, host='localhost', port=8321,
...
+        Returns:
+            divvy.Response, a namedtuple with these fields:
+                is_allowed: one of true, false, indicating whether quota was
+                    available.
+                current_credit: number of credit(s) available at the end of
+                    this command.
+                next_reset_seconds: time, in seconds, until credit next resets.
+        """
         cmd = self.translator.build_hit(**kwargs)
         self.connection.send(cmd)
```
Comment (aren) [inline divvy/client.py]: (line below) -- is it possible for recv() to only get partial results? does the memcached or redis client handle that case?

File: twisted_pool_poc.py (hunk 3 above)
Comment (RyanGWU82) [reply twisted_pool_poc.py]: Whoops, replaced with an `ArgumentParser` instead, now the hostname and port come from the command line.

File: twisted_poc.py (hunk 4 above)
Comment (RyanGWU82) [reply twisted_poc.py]: Same thing.

File: divvy/client.py (hunk 5 above)
Comment (RyanGWU82) [reply divvy/client.py]: The entire response can't be more than about 40 bytes: it's always `OK <bool> <int> <int> <int>`. Even with the largest ints, you're looking at approximately 41 bytes. The client is using a 1024-byte buffer so there's plenty of room in a single `recv()`.

The Redis client makes a single `socket.recv(1024)` so I'm pretty sure this is the right way to do it.

Comment (aren) [reply divvy/client.py]: Ok if redis does it that way I feel better about it. I was wondering if it was possible that a recv would only get (say) 20 bytes, and recv() would need to be called again one or more times to get the rest.

**Comment by aren** [comment]
Fix line 83 too

**Comment by RyanGWU82** [comment]
Closing this PR, will open a new one when the new Twisted implementation is ready.

**Pull request closed without merging**